Sources/NostrSDK/Events/Tags/EventTag.swift

```diff
@@ -150,11 +150,160 @@ public struct EventTag: RelayProviding, RelayURLValidating, Equatable {
+        guard eventTags.allSatisfy({ $0.marker == nil }) && eventTags.count >= 3 else {
+            return []
+        }
+
+        // The first event tag is the root and the last event tag is the one being replied to.
+        // Everything else in between is a mention.
+        return eventTags.dropFirst().dropLast()
+    }
+}
+
+/// Builds tags on a threaded event.
+/// See [NIP-10](https://github.com/nostr-protocol/nips/blob/master/10.md).
+public protocol ThreadedEventTagBuilding: NostrEventBuilding, RelayURLValidating {}
+public extension ThreadedEventTagBuilding {
+    /// Sets the ``ThreadedEventTagInterpreting`` event that is being replied to from this event that is being built.
+    /// See [NIP-10](https://github.com/nostr-protocol/nips/blob/master/10.md).
+    @discardableResult
+    func repliedEvent(_ repliedEvent: ThreadedEventTagInterpreting, relayURL: URL? = nil) throws -> Self {
+        let validatedRelayURL: URL?
+        if let relayURL {
+            validatedRelayURL = try validateRelayURL(relayURL)
+        } else {
+            validatedRelayURL = nil
+        }
+
+        if let rootEventTag = repliedEvent.rootEventTag {
+            // Maximize backwards compatibility with NIP-10 deprecated positional event tags
+            // by ensuring ordering of types of event tags.
+
+            // Root tag comes first.
+            if rootEventTag.marker == .root {
+                insertTags(rootEventTag.tag, at: 0)
+            } else {
+                // Recreate the event tag with a root marker if the one being read does not have a marker.
+                let rootEventTagWithMarker = try EventTag(eventId: rootEventTag.eventId, relayURL: rootEventTag.relayURL, marker: .root, pubkey: rootEventTag.pubkey)
+                insertTags(rootEventTagWithMarker.tag, at: 0)
+            }
+
+            // Reply tag comes last.
+            appendTags(try EventTag(eventId: repliedEvent.id, relayURL: validatedRelayURL, marker: .reply, pubkey: repliedEvent.pubkey).tag)
+        } else {
+            // If the event being replied to has no root marker event tag,
+            // the event being replied to is the root.
+            insertTags(try EventTag(eventId: repliedEvent.id, relayURL: validatedRelayURL, marker: .root, pubkey: repliedEvent.pubkey).tag, at: 0)
+        }
+
+        // When replying to a text event E, the reply event's "p" tags should contain all of E's "p" tags as well as the "pubkey" of the event being replied to.
+        // Example: Given a text event authored by a1 with "p" tags [p1, p2, p3] then the "p" tags of the reply should be [a1, p1, p2, p3] in no particular order.
+        appendTags(contentsOf: repliedEvent.tags.filter { $0.name == TagName.pubkey.rawValue })
+
+        // Add the author "p" tag if it was not already added.
+        if !tags.contains(where: { $0.name == TagName.pubkey.rawValue && $0.value == repliedEvent.pubkey }) {
+            if let validatedRelayURL {
```

Comment:
This specific logic around adding a relay to the p-tag is new. Which is why `TextNoteEventTests` was updated to match the assertion. All the other logic here was copied from `TextNoteEvent`.